java/src/main/java/com/google/protobuf/ZeroCopyEncoder.java

```diff
@@ -0,0 +1,552 @@
+      handler.onDataEncoded(value, offset, length, false);
+      return;
+    }
+
+    int capacity = buffer.remaining();
+    if (capacity >= length) {
+      // The entire value can be buffered.
+      buffer.put(value, offset, length);
+      return;
+    }
+
+    // It's a small buffer that won't completely fit into the buffer. Write it to the buffer,
+    // flushing as necessary.
+    int end = offset + length;
+    while (offset < end) {
+      if (capacity == 0) {
+        // The buffer is full, flush it.
+        flush();
+        capacity = buffer.remaining();
+      }
+
+      // Determine the number of bytes that can be written before another flush is required.
+      int bytesToWrite = min(end - offset, capacity);
+
+      buffer.put(value, offset, bytesToWrite);
+      offset += bytesToWrite;
+      capacity -= bytesToWrite;
+    }
+  }
+
+  private void writeRawBytes(ByteString value, int offset, int length) throws IOException {
+    if (length > buffer.capacity()) {
+      // It's a large buffer. Flush any currently buffered data and write out the value
+      // to the handler directly.
+      flush();
+
+      if (value instanceof LiteralByteString) {
+        LiteralByteString lbs = (LiteralByteString) value;
+        handler.onDataEncoded(lbs.bytes, lbs.getOffsetIntoBytes() + offset, length, false);
+      } else if (value instanceof ByteBufferByteString) {
+        ByteBuffer buf = ((ByteBufferByteString) value).buffer().asReadOnlyBuffer();
+        buf.position(offset);
+        buf.limit(offset + length);
+        handler.onDataEncoded(buf, false);
+      } else {
+        for (ByteBuffer buf : value.substring(offset, offset + length).asReadOnlyByteBufferList()) {
+          handler.onDataEncoded(buf, false);
+        }
+      }
+      return;
+    }
+
+    int capacity = buffer.remaining();
+
+    // If the buffer will hold the entire value, just add it to the buffer.
+    if (capacity >= length) {
+      value.copyTo(buffer, offset, length);
+      return;
+    }
+
+    // Write enough bytes to fill the buffer.
+    value.copyTo(buffer, offset, capacity);
+    offset += capacity;
+    length -= capacity;
+
+    // Flush the content of the buffer to the handler.
+    flush();
+
+    // We now have space for the remaining data.
+    value.copyTo(buffer, offset, length);
+    buffer.position(buffer.position() + length);
+  }
+
+  private void writeRawVarint32(int value) {
+    while (true) {
+      if ((value & ~0x7F) == 0) {
+        buffer.put((byte) value);
+        return;
+      }
+
+      buffer.put((byte) ((value & 0x7F) | 0x80));
+      value >>>= 7;
+    }
+  }
+
+  private void writeRawVarint64(long value) throws IOException {
+    while (true) {
+      if ((value & ~0x7FL) == 0) {
+        buffer.put((byte) value);
+        return;
+      }
+
+      buffer.put((byte) ((value & 0x7F) | 0x80));
+      value >>>= 7;
+    }
+  }
```

Comment:
Yes they are identical (except for the parameter type). The reason there are two is (IIUC) to correctly handle sign extension from int->long.  The only method that uses sign extension is writeInt32NoTag. If we were to alway call the 64-bit version, casting int->long, then every value written out would be an int64.